### DISCUSSION TITLE
Soft body dynamics for medical interactive simulation

### DISCUSSION THREAD

**Discussion by davide445** [mfem/mfem#2749] [Q&A] [answered]
Was considering MFEM for a medical interactive simulation project.
Searching a bit I didn't find any usage related to this sector, so wanted to ask how suitable are current solvers for soft body (human tissue) dynamics realtime simulation.
This will need to be integrated with biofluids simulation and perfusion, haptic feedback and immersive visualization (will be used a library with all compute based rendering pipeline to minimize latency).
There are more specialized frameworks such as SOFA, but we are targeting medical grade accurancy and AMD acelerators that are more in line with MFEM strengths.

**Comment by v-dobrev** [comment, marked as answer]
Hello Davide,

As far as I know, soft body tissue models are similar to nearly incompressible non-linear elasticity. If that is indeed the case, we have two examples that might be of interest for you to look at: `ex10`/`ex10p` (time-dependent Neo-Hookean non-linear elasticity) and `ex19`/`ex19p` (steady-state incompressible nonlinear elasticity) -- you can see short summaries of these and other examples/miniapps at https://mfem.org/examples.

Porting and optimizing these to GPUs will require some work, however, the elasticity kernels are of interest to other projects and I think we'll have initial support for elasticity on GPUs very soon. Also, such kernels are very similar to the TMOP (mesh optimization) kernels we have already ported to GPUs.

Regarding the integration with the other components you mentioned (like biofluids simulation and perfusion), do you envision using MFEM for those too or you have other tool(s) in mind? This sounds like a coupled fluid-structure interaction simulation which is also of interest to us, even though we currently do not have an example or miniapp like this.

Other project members can probably add to my summary: @tzanio, @jandrej, @jamiebramwell.

-Veselin

**Reply by davide445** [reply]
Hello Veselin
thanks for your answer.
We are discussing so far with OpenLB team members for the fluid sim, even if probably we will need to add SPH solver for the free flowing part outside veins/arteries.
Apart technical details I wanted to know also if there is any direct way to interact. I will need to decide soon what direction to follow on, since the tissue sim it's a key component driving all the other elements. SOFA team is based in EU that will make things easier for us, but not sure will be the best fit in term of sim accuracy and integration with AMD accelerators.
MFEM seems really good with the ability to model changing geometries (surgery will mean cutting), but also concerned about the suitability in more constrained hw than a supercomputer (for interactive sim the max will be a single workstation to avoid latencies and noise of a distributed compute server farm, so 4x accelerators) and the possibility to manage volumetric data (surgery will need something like RGB volumes to be visualized).

**Reply by BBBBBbruce** [reply]
Hello Davide, 
Can I ask if you and your team find a good solution for the realtime simulation for human tissue? I am working on this area as well. I think the pure computational mechanics method is not suitable in terms of realtime, especially for haptics. 

There are a couple of possible areas you can possibly look at:
1. Deep learning method -- still using FEM, but approximate it.
2. set different levels of details for the simulation -- FEM is calculating too much information, if you could identify the unnecessary and get rid of this part of simulation would accelerate a lot.
3. Some graphics methods, projective dynamics for example, but I am not sure how accurate you trying to preserve. There are always trade-offs between speed and accuracy.

**Reply by davide445** [reply]
Hello
we are prototyping with various methodologies, included ML.
As you wrote approximation is always useful to speedup things, the point for me is expand the computational capability of your architecture instead of compromising the quality.

**Reply by BBBBBbruce** [reply]
Yes, I agree, when the thing comes to the Biomedical area, quality is always the top priority. Have you tried PhysX 5.0? They have GPU supported Realtime FEM implementation. It might be useful.

**Reply by termi-official** [reply]
Since I am working on speeding up mechanics simulations, let me share my experience here. Projective dynamics has a rather limited capability - Tiantian Liu has developed a clever Quasi-Newton method (linear elasticity surrogate), but it also comes with the issue of how to costruct a suitable Jacobian, especially for advanced material models. For real time your best bet may be physics-informed neural networks (or in general techniques from the PDE-constrained zoo of technique) to derive a surrogate, unless you have really small number of elements and simple material models. You can use some offline FEM simulations to generate your training data.

**Reply by davide445** [reply]
> For real time your best bet may be physics-informed neural networks

Of course this is a path we are exploring

> You can use some offline FEM simulations to generate your training data.

The problem here is (apart the key scientific and medical aspects to setup a realistic enough simulation) as always the computational effort and elapsed time to generate a detailed training set for an high resolution simulation. This is one of the reasons I'm interested into MFEM.

**Reply by termi-official** [reply]
Great, I wish you good luck with this appraoch! 

Generating the data is ideed an issue. However, PINN's *seem* to have a very interesting property here. In comparison with learning decision making or vision problems, experiments which suggest that we do not need a super large data set to obtain a good set of weights (because we impose much structure through the PDE constraint). Note that the Networks also tend to be smaller, which might play a role. But I want to note that I personally do not have experience with PINN's beyond fitting surrogates for toy problems.
